File: basic-calculator/Solution224.java
// https://leetcode.com/problems/basic-calculator/

import java.util.Stack;

class Solution224 {

    // T: O(n), S: O(n)
    public int calculate(String exp) {
        int n = exp.length();
        var st = new Stack<Integer>();
        int num = 0, result = 0, sign = 1;
        for (char ch : exp.toCharArray()) {
            if (Character.isDigit(ch)) {
                int digit = ch - '0';
                num = num * 10 + digit;
            } else if (ch == '+') {
                result += num * sign;
                sign = 1;
                num = 0;
            } else if (ch == '-') {
                result += num * sign;
                sign = -1;
                num = 0;
            } else if (ch == '(') {
                st.push(result);
                st.push(sign);
                result = 0;
                sign = 1;
            } else if (ch == ')') {
                result += num * sign;
                result *= st.pop();
                result += st.pop();
                num = 0;
            }
        }
        result += num * sign;
        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution224();
        var s1 = "1 + 1";
        var s2 = " 2-1 + 2 ";
        var s3 = "(1+(4+5+2)-3)+(6+8)";
        System.out.println(sol.calculate(s1)); // 2
        System.out.println(sol.calculate(s2)); // 3
        System.out.println(sol.calculate(s3)); // 23
    }

}

File: basic-calculator/solution224.py
# https://leetcode.com/problems/basic-calculator/

class Solution224:

    # T: O(n), S: O(n)
    def calculate(self, s: str) -> int:
        st = []
        num, result, sign = 0, 0, 1

        for ch in s:
            if ch.isdigit():
                digit = int(ch)
                num = num * 10 + digit
            elif ch == '+':
                result += num * sign
                sign = 1
                num = 0
            elif ch == '-':
                result += num * sign
                sign = -1
                num = 0
            elif ch == '(':
                st.append(result)
                st.append(sign)
                result = 0
                sign = 1
            elif ch == ')':
                result += num * sign
                result *= st.pop()
                result += st.pop()
                num = 0

        result += num * sign
        return result


def main() -> None:
    sol = Solution224()
    s1 = "1 + 1"
    s2 = " 2-1 + 2 "
    s3 = "(1+(4+5+2)-3)+(6+8)"
    print(sol.calculate(s1))  # 2
    print(sol.calculate(s2))  # 3
    print(sol.calculate(s3))  # 23


if __name__ == "__main__":
    main()
